### ISSUE TITLE
wkhtmltoimage file sizes are ridiculous

### ISSUE THREAD

**Issue by achuinard** [wkhtmltopdf/wkhtmltopdf#2608]
If I generate a document with wkhtmltopdf, it comes out to about 35kb. This same document, with wkhtmltoimage -f png, comes out to about 4.4MB. Furthermore, running the generated PNG into something like tinypng.com results in a PNG of 35kb. Why is wkhtmltoimage creating an image with such a large file size?

**Comment by cdokolas** [comment]
I got a ~950KB generated file that should be ~35KB!

**Comment by PhilterPaper** [comment]
It sounds like the PNG file may be poorly compressed, or even *un*compressed. You mentioned using tinypng -- keep in mind that tinypng uses **lossy compression** to get very small file sizes, while most PNG compression methods are non-lossy (and thus larger). 

You might try firing up an image processor such as GIMP, and resaving the image to see if it gets any smaller. You might also try converting to (saving as) GIF to try out other non-lossy compressors, and as JPEG to try a lossy compressor. If you don't see any major size reduction with non-lossy compression, there probably isn't anything that can be done with wkHTMLtoImage. If you _can_ get major improvements with non-lossy compression, wkHTMLtoImage might be improved.

**Comment by JanisE** [comment]
ZIP compresses wkhtmltoimage's 11 MB file to 70 KB. PNG can certainly do better than that! :D

**Comment by PhilterPaper** [comment]
So apparently, wkHTMLtoImage is not doing _any_ compression on its output. Depending on the internals, it might be easy or it might be difficult to add non-lossy compression to PNG output. Do GIF (non-lossy) and JPEG (lossy) already compress their output? If not, maybe the intent was for users to use some external utility to do the compression (i.e., don't load up the module with extras).

ZIP should be a non-lossy compression, so 11MB-to-70kB is certainly plausible (PNG may use a different algorithm).

**Comment by ststeiger** [comment]
That's because you set quality to 100%, or 96%. 
If you create a PNG, you can without any problem set the quality to 50. 
If I set it to 100, I get a filesize of 150 MB. 
If I set it to 50, I get 729 KB. 
If I open the 150 MB file in pain, and save as PNG, I get 755KB. 
I can't see any quality difference by my-human-eyes on a quick glance when zooming-in. 

`--quality | <int> | Output image quality (between 0 and 100) (default 94)`

**Comment by cptjazz** [comment]
For PNGs it seems that the object streams in the PDF are *uncompressed*.  So a (losslessly) compressed PNG is blown up to huge raw bitmaps when copying it into the PDF objectstream.
For JPGs a DCTDecode filter is used -- basically the (compressed) JPG is copied into the PDF.

I would have expected that PNG streams are at least compressed losslessly (FlateDecode, LZWDecode, ...) because I have set *useCompression = true* in wkhtmltopdf.

**Comment by ststeiger** [comment]
Exactly, PNG is lossless, so you get the full-size bitmap-image @16 million colors, if you set quality to 100%. 
Compression can only achive that much. Lowering the quality to 50 will allow the program to "compress", by removing unnecessary accuracy. If you reduce quality on PNG, you still get a much better image than if you select jpeg.

**Comment by cptjazz** [comment]
PNGs internally can (and do!) use lossless compression algorithms.
The object stream in the PDF is uncompressed. It should be compressed with a *lossless* algorithm. Much smaller but no information is lost.

For example, my 24KB PNG blows up the PDF to 830KB -- the PNG internally is losslessly compressed, the PDF object stream is uncompressed.

Your approach is to throw away image information which is not what I want.

**Comment by Jmuccigr** [comment]
To @cptjazz's point, the first line in the [wikipedia entry](https://en.wikipedia.org/wiki/Portable_Network_Graphics) for png is 

> Portable Network Graphics (PNG /ˈpɪŋ/[2][3]) is a raster graphics file format that supports lossless data compression

You *can* use lossy techniques, but that's not what png is for and certainly not what the typical user would expect.

**Comment by michael-schaefer-eu** [comment]
I tested the behaviour of the --quality flag with jpg and png as @ststeiger suggested and came to the result that JPG will compress with loss and PNG will compress lossless. Here is some data to prove that:

JPG 100% (195 kB)
![grafik](https://user-images.githubusercontent.com/9930093/35007307-8c0235d6-faf9-11e7-818e-0017cd170b69.png)

JPG 20% (46 kB)
![grafik](https://user-images.githubusercontent.com/9930093/35007377-ce739c0c-faf9-11e7-8e33-c51aeded0f07.png)

JPG 1% (31 kB)
![grafik](https://user-images.githubusercontent.com/9930093/35007409-ea087618-faf9-11e7-8a89-1444c063e32d.png)

PNG 100% (7,25 mB !!!)
![grafik](https://user-images.githubusercontent.com/9930093/35007452-1450c146-fafa-11e7-8f48-928f1fc37335.png)

PNG 20% (59 kB)
![grafik](https://user-images.githubusercontent.com/9930093/35007504-2e1382a8-fafa-11e7-92fd-b282c96cffa1.png)

PNG 1% (58 kB)
![grafik](https://user-images.githubusercontent.com/9930093/35007532-4155783a-fafa-11e7-9135-3976d1fa1384.png)

Also it takes much longer to decode (convert with wkhtmltoimage) and encode (render in browser) the 1% PNG than the 20% PNG, so it is optimal to use a quality around 20 with PNG.

**Comment by michael-schaefer-eu** [comment]
Also I think the original issue is solved and can be closed.

**Comment by Jmuccigr** [comment]
I'd suggest that this shows that the default for PNGs should be 20% quality (or so) and not 100%, rather than forcing the user to set that. Why would no compression be the default when compression is lossless (assuming the time involved is not prohibitive)?

**Comment by PhilterPaper** [comment]
> JPG will compress with loss and PNG will compress lossless

Um, well, yes, that is to be expected. JPEG always does lossy compression (except where you order it to do 100% quality), and PNG normally does lossless compression, as does PDF. Lossy compression throws away fine detail (bulky high frequency information), and is adequate for photographs, but should never be used for charts, diagrams, and other non-photo images with large expanses of single colors, and sharp boundaries between colors (such as the example above), as the artifacting will be very obvious.

As to why 100% quality PNG is so much larger than 100% JPEG, I can't tell you. Maybe there's something wrong with the compression, or 100% JPEG is actually slightly lossy. As an experiment, you might take a given JPEG (from a camera), convert it to PNG in an image processor such as GIMP, and try different qualities (compression amounts) on both. This might show whether it's something inherent in the image formats, or a bad compression algorithm in the program.

Don't forget that JPEG discards image data with every save. I once worked with someone who was complaining that his JPEG images were suffering bit rot, and deteriorating over time. It turned out that he was frequently viewing them in an image editor, and rather than _quitting_ the program when done, he was _saving_ the image again! Each time, the images got softer and softer.

**Comment by michael-schaefer-eu** [comment]
No, it is non-expectable behaviour when the flag name is 'quality' but for PNG it sets actually 'compression rate' (as most image editors name this feature). Also that is not mentioned in the help:

> --quality \<int\> Output image quality (between 0 and 100) (default 94)

I think thats the reason for this issue, so I'd suggest to make the help message more clear, like:
> Output image compression rate (between 0 and 100) (default 94). May have effect on image quality.

**Comment by Jmuccigr** [comment]
That seems reasonable. Could we get the default for PNG set a lot closer to 0 than 100 though?

**Comment by PhilterPaper** [comment]
So you're saying that the PNG 0-100 value is the inverse of the JPEG 0-100 value? 100 for JPEG is no compression (or at least, no quality loss), while for PNG 100 would be the smallest possible? Is it still lossy compression for PNG, or always non-lossy? If it's non-lossy, it will have NO effect on image quality (and there would be no reason to select different levels, as you would simply compress the maximum amount).

**Comment by Jmuccigr** [comment]
I think @michael-schaefer-eu made the point above: there's a trade-off between file size and decompression time.

**Comment by teran-mckinney** [comment]
wkhtmltopdf is obviously not using any PNG (lossless) compression. PNG is a gzip'ed bitmap.

```
 #$ wkhtmltoimage about.html about.png
Loading page (1/2)
Error: Failed to load file:///ipfs/QmWgho8W12oimr4owjksfmaSusyQLcZUC9dr3746ieTj4Z, with network status code 203 and http status code 0 - Error opening /ipfs/QmWgho8W12oimr4owjksfmaSusyQLcZUC9dr3746ieTj4Z: No such file or directory
Rendering (2/2)                                                    
Done                                                               
Exit with code 1 due to network error: ContentNotFoundError
 #$ du -h about.png 
7.9M    about.png
 #$ gzip -9 about.png 
 #$ du -h about.png.gz 
228K    about.png.gz
```

You should not get a smaller file gzipping a png.

As a workaround for now I imagine there should be some kind of pngoptim (like jpegoptim) that would decompress/decode and recompress the image. As it is now, what could be a really useful tool is almost useless because of this. I am betting this is a fairly simple fix. While there certainly are lossy optimizations, I think 100% quality with lossless behavior would be perfect for text pages.

**Comment by ststeiger** [comment]
Might I say: 

We ran into rather severe problems with wkhtml2X. 
Apparently, the Chromium-Embedded version in QT is quite old, and somehow also unstable (heisenbugs - sometimes working - sometimes not - amd for no apparent reason). 

Chrome & Chromium now have a headless mode, to which you can connect by CDP (chromium debug-protocol). 
So you can directly use a headless chrome to capture screenshots and generate PDFs. 

You can find more information here: 
https://developers.google.com/web/updates/2017/04/headless-chrome
and here: 
https://medium.com/@dschnr/using-headless-chrome-as-an-automated-screenshot-tool-4b07dffba79a

We're using it in a production application, and so far, we're much happier with the results than we were with wkhtml. 

Here's an example in C# and VB.NET. 
The C# version still has some dummy code in it, from testing and doing other things ;) 
https://github.com/ststeiger/ChromeDevTools/blob/master/source/SampleNet4/ConverterTest.cs
https://github.com/ststeiger/ChromeDevTools/blob/master/source/SampleNet4VB/TestConverter.vb

**Comment by michael-schaefer-eu** [comment]
@teran-mckinney you obviously did not read the full discussion here, but you came to the same conclusion: the `--quality` flag is not intuitive. Its because the flag name and the help description is wrong.
please read [my comment above](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2608#issuecomment-359806393) to understand this better :)

**Comment by sogerc1** [comment]
Generating uncompressed PNGs by default makes no sense when there is a lossless compression anyway! I was transforming C3 charts to png, included them into a spreadsheet and then sent it to the user via email. My server was choking. After adding the --quality 20 option the final mail's size dropped almost 100 times without any visible quality change. My C3 charts are mostly white pixels.
The default for PNGs should be changed to something more reasonable!

**Comment by PhilterPaper** [comment]
Is this using a lossy compression for PNG? That could be the reason for a "quality" setting for PNG, unless there is a tradeoff between loss**less** compression (and possibly decompression) _time_ and the resulting _space_ used. Some lazy programmer may have decided to use "quality" to cover both JPEG-style lossy compression (where 100% is maximum quality/minimum compression) and a PNG lossy compression (where 100% is minimum quality/maximum compression). Then there are "quality" levels for lossless compression, setting a tradeoff between time and space. The use of "quality" to cover different things between different programs is inexcusable (and even more so _within_ a program), but appears to have happened. At the least, it should all be clearly documented. Caveat Emptor.

**Comment by sogerc1** [comment]
@PhilterPaper if you're asking me, I can't tell you. I'm not an expert in the subject, TBH I didn't even know PNGs _can_ be uncompressed or lossy (so I never even looked for the option until it became a problem). To my eyes it doesn't look like there's any loss, all the lines are just as sharp with quality 20 as with the default.
I also think that with today's processing power there's no reason not to compress the PNGs except if maybe someone has thousands of images to deal with but then wouldn't file size also become an issue? Not only in an email, but for ex. if a browser has to present 1000 PNGs then it would take less time to download and decompress them compared to when it needs to download 1000 huge images. There may be a lot of factors in play here but defaults should cover the most common usecases and I doubt it would be to put a BMP inside a PNG.
Also also, I agree that the documentation should be updated, JPEGs have quality so at first glance I would've ignored the option even if I were to look for it. It's only after I ran a command like:
`convert fromwkhtml.png imagemagick-test.png`
that I started to look around what's going on and found this issue, because the imagemagick file was so much smaller.

**Comment by michael-schaefer-eu** [comment]
I try to sum it up one more time:

- `PNG` is **always lossless compression** (always same quality independent of compression rate)
- `JPG` is **always with loss compression** (worse quality with higher compression rate)
- yes, the `--quality` param is used for both _png compression_ and _jpg compression_ (which then influences jpg quality)
  - which is simply a wrong param name (because misleading, just read this issue which exists for years now :facepalm: )
  - and also not good documented (see my [previous comment](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2608#issuecomment-359806393))
  - I guess it was first only the JPG functionality implemented with the `--quality` param - which has been (wrongly) reused later for PNG

About the optimal PNG compression rates, see my [previous comment](https://github.com/wkhtmltopdf/wkhtmltopdf/issues/2608#issuecomment-358076141).

It simply makes no sense to use `--param 1` compression, because the file size will be nearly the same as with ca. `--param 20` (but that are not exact rates, they depend on the use case and should be tested out).

**Comment by arisboch** [comment]
Until this is solved, I'm gonna use pngout. It'll take a while, but it gives results.
